File: release-gates/src/test/java/com/tongsr/kaleido/release/SimilarityAuditCliTest.java
package com.tongsr.kaleido.release;

import static org.junit.Assert.assertEquals;

import java.util.List;
import org.junit.Test;

public final class SimilarityAuditCliTest {
    @Test
    public void detectsSubstantialCopiedExpression() throws Exception {
        var tokens = java.util.stream.IntStream.range(0, 80).mapToObj(index -> "token" + index).toList();
        var digest = java.security.MessageDigest.getInstance("SHA-256");
        var hex = java.util.HexFormat.of().formatHex(digest.digest(
                String.join("\u0000", tokens).getBytes(java.nio.charset.StandardCharsets.UTF_8)));
        var findings = SimilarityAuditCli.compare(
                List.of(new SimilarityAuditCli.Source("candidate", tokens, hex)),
                List.of(new SimilarityAuditCli.Source("upstream", tokens, hex)));
        assertEquals(1, findings.size());
    }

    @Test
    public void allowsUnrelatedImplementations() throws Exception {
        var left = java.util.stream.IntStream.range(0, 80).mapToObj(index -> "left" + index).toList();
        var right = java.util.stream.IntStream.range(0, 80).mapToObj(index -> "right" + index).toList();
        assertEquals(List.of(), SimilarityAuditCli.compare(
                List.of(new SimilarityAuditCli.Source("candidate", left, "left")),
                List.of(new SimilarityAuditCli.Source("upstream", right, "right"))));
    }
}

File: release-gates/src/test/kotlin/com/tongsr/kaleido/release/SimilarityAuditCliTest.kt
package com.tongsr.kaleido.release

import java.nio.charset.StandardCharsets
import java.security.MessageDigest
import java.util.HexFormat
import org.junit.Assert.assertEquals
import org.junit.Test

class SimilarityAuditCliTest {
    @Test
    fun detectsSubstantialCopiedExpression() {
        val tokens = (0 until 80).map { index -> "token$index" }
        val digest = MessageDigest.getInstance("SHA-256")
        val hex = HexFormat.of().formatHex(
            digest.digest(tokens.joinToString("\u0000").toByteArray(StandardCharsets.UTF_8)),
        )
        val findings = SimilarityAuditCli.compare(
            listOf(SimilarityAuditCli.Source("candidate", tokens, hex)),
            listOf(SimilarityAuditCli.Source("upstream", tokens, hex)),
        )
        assertEquals(1, findings.size)
    }

    @Test
    fun allowsUnrelatedImplementations() {
        val left = (0 until 80).map { index -> "left$index" }
        val right = (0 until 80).map { index -> "right$index" }
        assertEquals(
            emptyList<SimilarityAuditCli.Finding>(),
            SimilarityAuditCli.compare(
                listOf(SimilarityAuditCli.Source("candidate", left, "left")),
                listOf(SimilarityAuditCli.Source("upstream", right, "right")),
            ),
        )
    }
}
